cg/meta/workflow/microsalt.py

```diff
@@ -306,3 +349,96 @@ def get_case_id_from_case(self, unique_id: str) -> Tuple[str, None]:
+            case_id=case_id,
+            failed_samples=failed_samples,
+            samples=samples,
+            run_dir_path=run_dir_path,
+        )
+
+    def qc_case_check(
+        self, case_id: str, samples: List[Sample], failed_samples: Dict, run_dir_path: Path
+    ) -> bool:
+        """Perform the final QC check for a microbial case based on failed samples."""
+        qc_pass: bool = True
+
+        for sample in failed_samples:
+            # Check if negative control failed
+            if sample.control == "negative":
+                qc_pass = False
+            # Check if MWR sample failed
+            if sample.application_version.application.tag == MicrosaltAppTags.MWRNXTR003:
+                qc_pass = False
+
+        # Check if more than 10% of MWX samples failed
+        if len(failed_samples) / len(samples) > MicrosaltQC.QC_PERCENT_THRESHOLD_MWX:
+            qc_pass = False
+
+        if not qc_pass:
+            LOG.warning(
+                f"Case {case_id} failed QC, see {run_dir_path}/QC_done.txt for more information."
+            )
+        else:
+            LOG.info(f"Case {case_id} passed QC.")
+
+        self.create_qc_done_file(
+            run_dir_path=run_dir_path,
+            failed_samples=failed_samples,
+        )
+        return qc_pass
+
+    def create_qc_done_file(self, run_dir_path: Path, failed_samples: Dict) -> None:
+        """Creates a QC_done when a QC check is performed."""
+        with open(os.path.join(run_dir_path, "QC_done.txt"), "w") as file:
+            for sample_dict in failed_samples.items():
+                file.write(f"{sample_dict[0].internal_id}: {json.dumps(sample_dict[1])} \n")
+
+    def qc_sample_check(self, sample: Sample, failed_samples: Dict, qc_file: Dict) -> None:
+        """Perform a QC on a sample."""
+
+        if sample.control == "negative":
+            if not self.check_external_negative_control_sample(sample):
+                failed_samples[sample] = {
+                    "Passed QC Reads": self.check_external_negative_control_sample(sample)
+                }
+                LOG.warning(f"Negative control sample {sample.internal_id} failed QC.")
+        else:
+            reads_pass: bool = sample.sequencing_qc
+            coverage_10x_pass: bool = self.check_coverage_10x(sample.internal_id, qc_file)
+            if not reads_pass or not coverage_10x_pass:
+                failed_samples[sample] = {"Passed QC Reads": reads_pass}
+                failed_samples[sample]["Passed Coverage 10X"] = coverage_10x_pass
+                LOG.warning(f"Sample {sample.internal_id} failed QC.")
+
+    def check_coverage_10x(self, sample_name: str, qc_file: Dict) -> bool:
+        """Check if a sample passed the coverage_10x criteria."""
+        try:
+            return (
+                qc_file[sample_name]["microsalt_samtools_stats"]["coverage_10x"]
+                >= MicrosaltQC.COVERAGE_10X_THRESHOLD
+            )
+        except Exception as e:
```

Comment:
Could you specify `KeyError` here?

Reply:
Catching all exceptions is considered bad practice (https://stackoverflow.com/questions/10594113/bad-idea-to-catch-all-exceptions-in-python). Try to only except the error/s you expect might occur, in this case `TypeError`, right?

Reply:
Great point, I will address this!